### DISCUSSION TITLE
DLL export

### DISCUSSION THREAD

**Discussion by jankrassnigg** [jrouwe/JoltPhysics#503] [Q&A]
In our engine Jolt is integrated as a plugin. Currently that plugin statically links against the Jolt library. Game code typically resides in its own plugin. However, when somebody wants to access Jolt directly, for instance to write their own character controller, they need to link against the Jolt library as well. Since that is a static library atm, that means we end up with some stuff duplicated and ultimately things just crash.

Usually the solution is to mark all classes as __declspec(dllexport) and turn the Jolt library into a DLL, than multiple other DLLs can link to it, everything gets shared and nothing gets duplicated.

If I were to add this, I would keep the default as it is, ie. use a macro that is by default a no-op, but that can be enabled to do DLL export. This would be quite a large change, since it would affect all classes, global variables, etc.

I was wondering whether you would accept such a change? I haven't started on anything and I don't know when I would be able to invest that time, but wanted to know your opinion on this.

**Comment by jrouwe** [comment]
Personally I'm not a big fan of DLL exports for C++ code as my experience with it is that it quickly gets out of control. Where do you draw the line? E.g. does Vec3 need to be exported too? Does that mean that almost every function call on a Vec3 involves calling across the DLL boundary?  Also changes to any of the APIs are harder afterwards as it will break the DLL interface too and may silently cause pieces of compiled code to fail. Usually a C style DLL interface is much cleaner and there are currently 2 people working on such interfaces (see main page), so I was hoping to not have to provide a C++ style DLL interface.

IMO, a cleaner solution in your case would be to create wrappers for functionality that game libraries may require. That would also mean that if ever you decide to switch from Jolt to PhysicsLibraryX that you only need to update the wrappers and that game code should continue to work.

**Reply by jankrassnigg** [reply]
All you need to do, is to add a single macro to each class that belongs to the public interface. E.g.:

```
class Body
{
    void Func1();
    void Func2();
    static int globalValue;
};
```

becomes

```
class JPH_EXPORT Body
{
    void Func1();
    void Func2();
    static int globalValue;
};
```

In my opinion that's not much of a hassle.

There is another option, where you don't put it next to the class name, but in front of every single member function, which is much more messy. Maybe you've seen that in some libraries. That way gives you more control over what exactly gets exported so you can trim down the list. In my experience this is rarely needed.

> Where do you draw the line? E.g. does Vec3 need to be exported too?

Depends. If it has out-of-line functions functions, then yes. If it is completely implemented as inline (or template) functions, then no.

> Does that mean that almost every function call on a Vec3 involves calling across the DLL boundary?

Absolutely not. Inlining still works exactly the same. If the compiler can see the implementation, it will inline things. Performance generally isn't affected much by it on x86 and on other platforms you would still use static linking. PhysX is also a DLL on Windows, and it was never a problem for them.

> Also changes to any of the APIs are harder afterwards as it will break the DLL interface too and may silently cause pieces of compiled code to fail.

I disagree with this conceptually. This can only happen if you swap out the DLL but don't compile the user code again. The only use case that I can think of here, is mod support for a released game. In such a situation you have to be very, very careful in general, how you expose your interfaces. In almost all other cases when you update Jolt, you would also compile all user code again. And I don't mean "force clean and recompile all", I mean just regular "hit compile and let the compiler decide what needs to be done", because it will see which headers in Jolt have changed. In such a case nothing can silently fail (disregarding compiler bugs of course).

> Usually a C style DLL interface is much cleaner
> IMO, a cleaner solution in your case would be to create wrappers for functionality that game libraries may require. That would also mean that if ever you decide to switch from Jolt to PhysicsLibraryX that you only need to update the wrappers and that game code should continue to work.

Both a C style interface and a custom wrapper have some serious issues. A big one is that they are easily out of date. Let's say you make a cool new vehicle feature, someone who uses our engine wants to use that. With DLL export all they (or we) need to do is to update Jolt and then they can use it right away. With any wrapper, we would first need to update that as well. Every API change that you make, leads to extra work. This is simply out of scope for our tiny team.

A custom wrapper is even worse. Unreal had that. To my knowledge they recently got rid of it. A while ago I looked at it, and it was littered with ifdef PHYSX and ifdef CHAOS etc. The problem with physics libraries is, that they are not easily abstractable. Just compare all the options for constraints in PhysX and Jolt. They are entirely different. You just can't map how Jolt handles stiffness, motors and such to how PhysX does it. The idea that you just replace the implementation and have things work the same way is not realistic. This works with graphics APIs, because they ultimately map to the same hardware and thus have the same capabilities, but physics engines are only similar on what problem they try to solve, they diverge massively when it comes to how exactly they solve them, and thus settings that lead to a stable simulation in one engine, are simply not available in another or result in complete garbage.

Generally in our engine we try to provide as much as possible for "day-to-day" use, so that most people don't need direct access to Jolt. However, for advanced uses, we want them to be able to use Jolt directly, so that they have access to every single detail and can make their use case work. You would never switch out the physics engine afterwards in such a scenario. You also wouldn't want to be forced to use a C API.

Maybe I'll do the DLL export stuff and just open a PR and let you have a look at it. I don't think it will be as bad as you think.

**Reply by jrouwe** [reply]
Ok, show me what it will look like.

About Vec3: It would be possible not to export classes like this at all right? Even if they have definitions in the cpp file that would just mean that there is a version of that function in the DLL and a version of the function in client code?

**Reply by jankrassnigg** [reply]
PR is here: https://github.com/jrouwe/JoltPhysics/pull/513

Regarding your Vec3 question: no, a function that is not inline will not be visible to the linker (like in a static library), when you compile something as a DLL. So either it is marked for export and then it will always call into the DLL, or it is not marked for export, but then trying to call it will result in a linker error that there is a missing symbol.

I don't know how familiar you are with how static and dynamic libraries differ during linking. It's a topic that took me quite a while to really grasp in detail. If you like, I can write a full explanation.

**Reply by jankrassnigg** [reply]
Thanks for getting this in. It will be very useful to us.
